src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,642 @@
+#include "librbd/ExclusiveLock.h"
+#include "librbd/MirroringWatcher.h"
+#include "journal/Settings.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "cls/journal/cls_journal_client.h"
+#include "librbd/image/RemoveRequest.h"
+#include "librbd/journal/RemoveRequest.h"
+#include "librbd/operation/TrimRequest.h"
+#include "librbd/mirror/DisableRequest.h"
+#include "librbd/operation/DisableFeaturesRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::image::RemoveRequest: " << this << " " \
+                           << __func__ << " "
+
+namespace librbd {
+namespace image {
+
+using util::create_context_callback;
+using util::create_rados_ack_callback;
+
+template<typename I>
+RemoveRequest<I>::RemoveRequest(IoCtx &ioctx, const std::string &image_name, const std::string &image_id,
+                                bool force, ProgressContext &prog_ctx, ContextWQ *op_work_queue,
+                                Context *on_finish) :
+  m_image_name(image_name), m_image_id(image_id), m_force(force),
+  m_prog_ctx(prog_ctx), m_op_work_queue(op_work_queue), m_on_finish(on_finish) {
+  m_ioctx.dup(ioctx);
+  m_cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+
+  m_image_ctx = new I((m_image_id.empty() ? m_image_name : std::string()),
+                      m_image_id, nullptr, m_ioctx, false);
+}
+
+template<typename I>
+void RemoveRequest<I>::send() {
+  ldout(m_cct, 20) << dendl;
+
+  open_image();
+}
+
+template<typename I>
+void RemoveRequest<I>::open_image() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_open_image>(this);
+
+  m_image_ctx->state->open(ctx);
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_open_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
+    m_retval = *result;
+    mirror_image_remove();
+    return nullptr;
+  }
+
+  m_image_id = m_image_ctx->id;
+  m_image_name = m_image_ctx->name;
+  m_header_oid = m_image_ctx->header_oid;
+  m_old_format = m_image_ctx->old_format;
+  m_unknown_format = false;
+
+  check_exclusive_lock();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::mirror_image_remove() {
+  ldout(m_cct, 20) << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::mirror_image_remove(&op, m_image_ctx->id);
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_mirror_image_remove>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, rados_completion, &op);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_mirror_image_remove(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if ((*result < 0) && (*result != -ENOENT)) {
```

Comment:
Please look at the original code. We need to check for -EOPNOTSUPP too, just in case we are running on an older cluster. Also, I don't like very much unnecessary extra parenthesis.